Mathlib/Combinatorics/SetFamily/KruskalKatona.lean

```diff
@@ -0,0 +1,344 @@
+  · rw [← hU'] at same_size
+    have : erase U x = ∅ := by rw [← Finset.card_eq_zero, card_erase_of_mem Hx, ← hU']
+    have : erase V (min' V hV) = ∅ := by
+      rw [← Finset.card_eq_zero, card_erase_of_mem (min'_mem _ _), ← same_size]
+    rw [‹erase U x = ∅›, ‹erase V (min' V hV) = ∅›]
+    exact isCompressed_self _ _
+  refine h₂ ⟨UVd.mono (erase_subset ..) (erase_subset ..), ?_, ?_, ?_, ?_⟩ (card_erase_lt_of_mem Hx)
+  · rw [card_erase_of_mem (min'_mem _ _), card_erase_of_mem Hx, same_size]
+  · rwa [← card_pos, card_erase_of_mem Hx, tsub_pos_iff_lt]
+  · rwa [← Finset.card_pos, card_erase_of_mem (min'_mem _ _), ← same_size, tsub_pos_iff_lt]
+  · exact (Finset.max'_subset _ $ erase_subset _ _).trans_lt (max_lt.trans_le $
+      le_max' _ _ $ mem_erase.2 ⟨(min'_lt_max'_of_card _ (by rwa [← same_size])).ne', max'_mem _ _⟩)
+
+/-- If we're compressed by all useful compressions, then we're an initial segment. This is the other
+key Kruskal-Katona part. -/
+lemma isInitSeg_of_compressed {ℬ : Finset (Finset α)} {r : ℕ} (h₁ : (ℬ : Set (Finset α)).Sized r)
+    (h₂ : ∀ U V, UsefulCompression U V → IsCompressed U V ℬ) : IsInitSeg ℬ r := by
+  refine ⟨h₁, ?_⟩
+  rintro A B hA ⟨hBA, sizeA⟩
+  by_contra hB
+  have hAB : A ≠ B := ne_of_mem_of_not_mem hA hB
+  have hAB' : A.card = B.card := (h₁ hA).trans sizeA.symm
+  have hU : (A \ B).Nonempty := sdiff_nonempty.2 fun h ↦ hAB $ eq_of_subset_of_card_le h hAB'.ge
+  have hV : (B \ A).Nonempty :=
+    sdiff_nonempty.2 fun h ↦ hAB.symm $ eq_of_subset_of_card_le h hAB'.le
+  have disj : Disjoint (B \ A) (A \ B) := disjoint_sdiff.mono_left sdiff_subset
+  have smaller : max' _ hV < max' _ hU := by
+    obtain hlt | heq | hgt := lt_trichotomy (max' _ hU) (max' _ hV)
+    · rw [← compress_sdiff_sdiff A B] at hAB hBA
+      cases hBA.not_lt $ toColex_compress_lt_toColex hlt hAB
+    · exact (disjoint_right.1 disj (max'_mem _ hU) $ heq.symm ▸ max'_mem _ _).elim
+    · assumption
+  refine hB ?_
+  rw [← (h₂ _ _ ⟨disj, card_sdiff_comm hAB'.symm, hV, hU, smaller⟩).eq]
+  exact mem_compression.2 (Or.inr ⟨hB, A, hA, compress_sdiff_sdiff _ _⟩)
+
+attribute [-instance] Fintype.decidableForallFintype
+
+-- TODO: There's currently a diamond
+-- import Mathlib.Order.Fin.Basic
+-- example (n : ℕ) : instDecidableEqFin n = instDecidableEq_mathlib := rfl
+attribute [-instance] instDecidableEqFin
+
+/-- This measures roughly how compressed the family is.
+
+Note that this does depend on the order of the ground set, unlike the Kruskal-Katona theorem itself
+(although `kruskal_katona` currently is stated in an order-dependent manner). -/
+private def familyMeasure (𝒜 : Finset (Finset (Fin n))) : ℕ := ∑ A in 𝒜, ∑ a in A, 2 ^ (a : ℕ)
+
+/-- Applying a compression strictly decreases the measure. This helps show that "compress until we
+can't any more" is a terminating process. -/
+lemma familyMeasure_compression_lt_familyMeasure {U V : Finset (Fin n)} {hU : U.Nonempty}
```

Comment:
private, probably? familyMeasure is private  and `kruskal_katona_helper` is too

Reply:
Rah yes I always assume that lemmas whose type include a private definition are themselves private, but this seems not to be the case.

Reply:
I have opened https://github.com/leanprover/lean4/issues/5002 to possibly fix this confusing behavior